File: src/execgo_playground/scenarios/loader.py
from __future__ import annotations

import importlib
from pathlib import Path
from typing import Any, Callable

from ..models import PromptPack, ScenarioExpected, ScenarioSpec, VerifierResult
from ..utils import read_json, repo_root


def load_scenario(scenario_id: str) -> ScenarioSpec:
    base = repo_root() / "scenarios" / scenario_id
    raw = read_json(base / "scenario.json")
    expected = read_json(base / "expected.json")
    prompt_dir = base / "prompt_pack"
    prompt_pack = PromptPack(
        system_prompt=(prompt_dir / "system.txt").read_text(encoding="utf-8").strip(),
        user_prompt=(prompt_dir / "user.txt").read_text(encoding="utf-8").strip(),
        constraints_prompt=(prompt_dir / "constraints.txt").read_text(encoding="utf-8").strip(),
    )
    return ScenarioSpec(
        id=raw["id"],
        description=raw["description"],
        seed=raw.get("seed", 7),
        input=raw.get("input", {}),
        prompt_pack=prompt_pack,
        expected=ScenarioExpected(checks=expected),
        verifier_ref=raw["verifier_ref"],
        allowed_chaos=raw.get("allowed_chaos", []),
        fixtures_dir=str(base / "fixtures"),
        scenario_dir=str(base),
    )


def load_verifier(ref: str) -> Callable[[ScenarioSpec, dict[str, Any]], VerifierResult]:
    module_name, attr = ref.split(":", maxsplit=1)
    module = importlib.import_module(module_name)
    verifier = getattr(module, attr)
    return verifier

File: src/execgo_playground/scenarios/loader.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import {
  type PromptPack,
  type ScenarioSpec,
  type VerifierResult,
  objectValue,
} from "../models.js";
import { readJson, repoRoot } from "../utils.js";
import {
  verifyCodegenExec,
  verifyLongChainDag,
  verifyMultiStepAgent,
  verifyVulnScan,
} from "./verifiers.js";

export async function loadScenario(scenarioId: string): Promise<ScenarioSpec> {
  const base = path.join(repoRoot(), "scenarios", scenarioId);
  const raw = objectValue(await readJson(path.join(base, "scenario.json")));
  const expected = objectValue(await readJson(path.join(base, "expected.json")));
  const promptDir = path.join(base, "prompt_pack");
  const promptPack: PromptPack = {
    system_prompt: (await readFile(path.join(promptDir, "system.txt"), "utf8")).trim(),
    user_prompt: (await readFile(path.join(promptDir, "user.txt"), "utf8")).trim(),
    constraints_prompt: (await readFile(path.join(promptDir, "constraints.txt"), "utf8")).trim(),
  };
  return {
    id: String(raw.id),
    description: String(raw.description),
    seed: raw.seed === undefined ? 7 : Number(raw.seed),
    input: objectValue(raw.input),
    prompt_pack: promptPack,
    expected: { checks: expected },
    verifier_ref: String(raw.verifier_ref),
    allowed_chaos: Array.isArray(raw.allowed_chaos) ? raw.allowed_chaos.map(String) : [],
    fixtures_dir: path.join(base, "fixtures"),
    scenario_dir: base,
  };
}

export type Verifier = (spec: ScenarioSpec, context: Record<string, unknown>) => Promise<VerifierResult> | VerifierResult;

export function loadVerifier(ref: string): Verifier {
  const name = ref.includes(":") ? ref.split(":").at(-1)! : ref;
  const catalog: Record<string, Verifier> = {
    verify_codegen_exec: verifyCodegenExec,
    verify_vuln_scan: verifyVulnScan,
    verify_multi_step_agent: verifyMultiStepAgent,
    verify_long_chain_dag: verifyLongChainDag,
    verifyCodegenExec,
    verifyVulnScan,
    verifyMultiStepAgent,
    verifyLongChainDag,
  };
  const verifier = catalog[name];
  if (!verifier) {
    throw new Error(`unknown verifier: ${ref}`);
  }
  return verifier;
}
